### ISSUE TITLE
Non thread safe exception handling for .NET

### ISSUE THREAD

**Issue by paulkulakov** [swig/swig#1383]
Hi all!

Seems like exception handling in .NET version is not thread safe.

The following fragment from csharphead.swg
```
public class SWIGPendingException {
    [global::System.ThreadStatic]
    private static global::System.Exception pendingException = null;
    private static int numExceptionsPending = 0;
...
```
should be changed to
```
public class SWIGPendingException {
    [global::System.ThreadStatic]
    private static global::System.Exception pendingException = null;
    [global::System.ThreadStatic]
    private static int numExceptionsPending = 0;
...
```
making numExceptionsPending thread-specific as well as pendingException

**Comment by paulkulakov** [comment]
Hi William,

OK, you are right. I didn't understand the idea of member numExceptionsPending.
The problem I faced with was using SWIG-based library (NQuantLib) in different .NET app domains: managed callbacks (ExceptionDelegate) are not called in the same app domain as parent mamanged call.
The diagram is:
```
Domain2: ManagedCall (SWIG .NET wrapper)
Unmanaged call which generates app exception
Domain1: exception callback (e.g. SetPendingApplicationException)
Domain2: SWIGPendingException.Pending which returns false instead of true
```

So I had to redesign my application to use single domain model.

Thanks.
